### PR TITLE
Add metadata registration macros

### PR DESCRIPTION
**Pull request by molpopgen**
- start on metadata registration macros
- add registration + rountrip tests for json
- test w/newtypes
- add remaining macros and tests
- restrict macro to feature set

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -45,11 +45,18 @@
 //!
 //! * `provenance`
 //!     * Enables [`provenance`]
-//! * `serde_json_metadata`
-//!     * Enables [`serde_json_metadata`] macro.
-//! * `serde_bincode_metadata`
-//!     * Enables [`serde_bincode_metadata`] macro.
+//! * `derive` enables the following procedural macros:
```
Comment (MomoLangenstein) [inline src/lib.rs:48]: I would call them `#[derive(…)]` macros (procedural is more general)

File: src/lib.rs [hunk 2]
```diff
@@ -69,6 +76,15 @@
 #[allow(deref_nullptr)]
 pub mod bindings;
 
+#[cfg(feature = "derive")]
+pub extern crate tskit_derive;
```
Comment (MomoLangenstein) [inline src/lib.rs:80]: You could probably make this `#[doc(hidden)]` as nobody needs to directly access the derive crate as you re-export all macros

File: src/lib.rs [hunk 3]
```diff
@@ -69,6 +76,15 @@
...
+
+#[cfg(feature = "derive")]
+pub use tskit_derive::{
```
Comment (MomoLangenstein) [inline src/lib.rs:83]: I’d re-export the macros in the same module as the traits, so if a user imports the trait they get both (this is a nifty import overload feature)

File: src/metadata.rs [hunk 4]
```diff
@@ -1,88 +1,178 @@
...
+//! Metadata refers to data that client code may need to associate
+//! with table rows, but the data are not necessary to perform algorithms
+//! on tables nor on trees.
+//!
+//! For complete details, see the data model descriptions
+//! [`here`](https://tskit.dev/tskit/docs/stable/)
+//!
+//! The most straightfoward way to implement metadata
+//! is to use the optional `derive` feature of `tskit`.
+//! This feature enables procedural macros to convert
```
Comment (MomoLangenstein) [inline src/metadata.rs:12]: Derive macros

File: src/metadata.rs [hunk 5]
```diff
@@ -1,88 +1,178 @@
...
+//! your types to metadata types via [`serde`](https://docs.rs/serde).
+//!
+//! Note that you will need to add `serde` as a dependency of your
+//! package, as you will need its `Serialize` and `Deserialize`
```
Comment (MomoLangenstein) [inline src/metadata.rs:16]: Derive macros

File: src/metadata.rs [hunk 6]
```diff
@@ -1,88 +1,178 @@
...
+//! procedural macros available.
+//!
+//! Without the procedural macros provided by tskit, you must `impl` [`MetadataRoundtrip`]
```
Comment (MomoLangenstein) [inline src/metadata.rs:19]: Derive macros

File: src/metadata.rs [hunk 7]
```diff
@@ -1,88 +1,178 @@
...
+//!     )
+//!     .unwrap();
+//!
+//! let decoded = tables.nodes().metadata::<Metadata>(id).unwrap().unwrap();
+//! assert_eq!(decoded.data, "Bananas".to_string());
+//! ```
+//!
+//! # Technial details and notes
+//!
+//! * The procedural macros currently support two `serde` methods:
```
Comment (MomoLangenstein) [inline src/metadata.rs:154]: Derive macros

File: src/metadata.rs [hunk 8]
```diff
@@ -1,88 +1,178 @@
...
+//!   `serde_json` and `bincode`.
+//! * A concept like "mutation metadata" is the combination of two traits:
+//!   [`MetadataRoundtrip`] plus [`MutationMetadata`].
+//!   The latter is a marker trait.
+//! * The procedual `derive` macro for mutation metadata
```
Comment (MomoLangenstein) [inline src/metadata.rs:159]: If you put both in the same namespace, they will overlap, which is the desired behaviour (e.g. in serde serde::Serialize refers to both the trait and the derive macro)

File: tskit-derive/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,105 @@
+use proc_macro::TokenStream;
+
+fn impl_serde_json_roundtrip(name: &syn::Ident) -> TokenStream {
+    let gen = quote::quote!(
```
Comment (MomoLangenstein) [inline tskit-derive/src/lib.rs:4]: What you might want to do is to prefix the absolute paths with ::, e.g. ::tskit or ::serde_json. That way, the paths won’t clash with local user modules of the same name but must refer to external crates.

File: tskit-derive/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+fn impl_metadata_roundtrip_macro(ast: &syn::DeriveInput) -> Result<TokenStream, syn::Error> {
+    let name = &ast.ident;
+    let attrs = &ast.attrs;
+
+    for attr in attrs.iter() {
+        if attr.path.is_ident("serializer") {
+            let lit: syn::LitStr = attr.parse_args().unwrap();
+            let serializer = lit.value();
+
+            if serializer == "serde_json".to_string() {
```
Comment (MomoLangenstein) [inline tskit-derive/src/lib.rs:62]: If &serializer == “…” avoids the allocation

File: tskit-derive/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+                return Ok(impl_serde_bincode_roundtrip(name));
+            } else {
+                proc_macro_error::abort!(serializer, "is not a supported protocol.");
+            }
+        } else {
+            proc_macro_error::abort!(attr.path, "is not a supported attribute.");
+        }
+    }
+
+    unimplemented!()
```
Comment (MomoLangenstein) [inline tskit-derive/src/lib.rs:74]: Abort_call_site with “we need a serializer attribute” or something

File: src/lib.rs (hunk 1 above)
Comment (molpopgen) [reply src/lib.rs:48]: done

File: src/lib.rs (hunk 2 above)
Comment (molpopgen) [reply src/lib.rs]: done

File: src/lib.rs (hunk 3 above)
Comment (molpopgen) [reply src/lib.rs]: done

File: src/metadata.rs (hunk 4 above)
Comment (molpopgen) [reply src/metadata.rs]: done

File: src/metadata.rs (hunk 5 above)
Comment (molpopgen) [reply src/metadata.rs:16]: done

File: src/metadata.rs (hunk 6 above)
Comment (molpopgen) [reply src/metadata.rs]: done

**Comment by molpopgen** [comment]
Thanks for the comments @MomoLangenstein.  I didn't know about that "magic" overloading.  I'd assumed it wouldn't work.

**Comment by MomoLangenstein** [comment]
Of course, always happy to help. I also didn’t actively know about them until very recently. This PR now definitely looks very good to me

**Comment by molpopgen** [comment]
> I also didn’t actively know about them until very recently.

Neat!  Gonna merge...

**Pull request merged**